src/test/ui/const-generics/macro_rules-braces.rs

```diff
@@ -0,0 +1,43 @@
+// revisions: full min
+#![cfg_attr(full, allow(incomplete_features))]
+#![cfg_attr(full, feature(const_generics))]
+#![cfg_attr(min, feature(min_const_generics))]
+
+fn test<const N: usize>() {
+    struct Foo<const M: usize>;
+    macro_rules! foo {
+        ($x:expr) => {
+            [u8; $x] //[full]~ ERROR constant expression depends
+        }
+    }
+    macro_rules! bar {
+        ($x:expr) => {
+            [u8; { $x }] //[full]~ ERROR constant expression depends
+        }
+    }
+    macro_rules! baz {
+        ( $x:expr) => {
+            Foo<$x> //[full]~ ERROR constant expression depends
+        }
+    }
+    macro_rules! biz {
+        ($x:expr) => {
+            Foo<{ $x }> //[full]~ ERROR constant expression depends
+        };
+    }
+    
+    let _: foo!(N);
+    let _: foo!({ N });
+    let _: foo!({{ N }}); //[min]~ ERROR generic parameters may not
+    let _: bar!(N);
+    let _: bar!({ N }); //[min]~ ERROR generic parameters may not
+    let _: baz!(N); //~ ERROR expressions must be enclosed in braces
```

Comment:
that one's surprising, `Foo<N>` should be allowed but apparently doesn't work inside of macros

Reply:
It would be nice to fix this.

Reply:
do you or @petrochenkov have an idea on where this might go wrong?

That's not really something I am familiar with

Reply:
I believe the issue is here:
https://github.com/rust-lang/rust/blob/ffe52882ed79be67344dd6085559e308241e7f60/compiler/rustc_parse/src/parser/diagnostics.rs#L1823
`expr_is_valid_const_arg` returns false for an identifier like `N`, because we usually simply parse identifiers as type arguments instead, and then resolve them to const arguments later. (The comment at the top of `handle_unambiguous_unbraced_const_arg` is misleading and should probably be changed: it doesn't just handle an error code, it is also for parsing valid const arguments that are not surrounded in braces.) I think if you add the identifier case to `expr_is_valid_const_arg`, this might work as expected. I'm not quite sure off the top of my head why this is the only place we're hitting this issue, though, or why there's not an ambiguity issue with type arguments.

We can do this as a follow up, though.